src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -549,31 +554,39 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
         return new
 
     def filter(self, col: Expression) -> "SelectStatement":
-        if self.flatten_disabled:
-            can_be_flattened = False
-        else:
-            dependent_columns = derive_dependent_columns(col)
-            can_be_flattened = can_clause_dependent_columns_flatten(
-                dependent_columns, self.column_states
-            )
+        dependent_columns = derive_dependent_columns(col)
+        can_be_flattened = (
+            not self.flatten_disabled
+        ) and can_clause_dependent_columns_flatten(
+            dependent_columns, self.column_states
+        )
         if can_be_flattened:
             new = copy(self)
             new.from_ = self.from_.to_subqueryable()
             new.pre_actions = new.from_.pre_actions
             new.post_actions = new.from_.post_actions
-            new.where = And(self.where, col) if self.where is not None else col
             new._column_states = self._column_states
+            if self.where:
```

Comment:
Core change #1: merge where clauses between parent query and child query by AND the condition expressions, track the dependent columns of a WHERE clause, merge two dependent columns sets using UNION.
@sfc-gh-sfan @sfc-gh-yixie